### PR TITLE
Add public_suffix host validation option

### PR DESCRIPTION
**Pull request by zapo**
This adds a public_suffix validation option allowing to validate URI host against https://publicsuffix.org/ database (adds https://github.com/weppos/publicsuffix-ruby as dependency).

### CODE REVIEW

**Comment by kritik** [comment]
Which urls are valid and invalid if public suffix is true or false

**Comment by zapo** [comment]
When setting public_suffix option false it just behaves normally (won't validate for public suffix). So something like `http://domain.unknown_tld` would be valid.

When setting public_suffix option true it validates that the url tld or suffix matches the public suffix db here https://publicsuffix.org/list/public_suffix_list.dat something like `http://domain.unknown_tld` would be invalid.

**Comment by zapo** [comment]
I can understand maintaining an additional dependency for this concern might not be appealing since this could be handled by some wrapping validator. Feel free to close if that's the case.

**Comment by kritik** [comment]
Can you give more examples, i.e.  public_suffix option true which domain is true

**Comment by fbernier** [comment]
@kritik it validates against the list of all valid TLDs as seen here: https://publicsuffix.org/

**Comment by kritik** [comment]
@ericraio Finally I checked when it's needed. I don't see any reason why it should be here but as there's a lot of wishers, then we can merge. Can you resolve conflicts?

**Comment by zapo** [comment]
Whew, revival of 4 years old PR hehe. I just updated it :) Tests seem to pass.

File: validate_url.gemspec [hunk 1]
```diff
@@ -39,15 +39,22 @@ Gem::Specification.new do |s|
 
     if Gem::Version.new(Gem::VERSION) >= Gem::Version.new('1.2.0') then
       s.add_runtime_dependency(%q<activemodel>, [">= 3.0.0"])
+      s.add_runtime_dependency(%q<public_suffix>, [">= 2.0.0"])
       s.add_development_dependency(%q<rspec>, [">= 3.0.0"])
       s.add_development_dependency(%q<diff-lcs>, [">= 1.1.2"])
+      s.add_development_dependency(%q<rake>)
```
Comment (zapo) [inline validate_url.gemspec]: This might not be needed.

Comment (kritik) [reply validate_url.gemspec]: @zapo yes, remove it

Comment (zapo) [reply validate_url.gemspec]: Fixed.

File: lib/validate_url.rb [hunk 2]
```diff
@@ -19,8 +21,10 @@ def validate_each(record, attribute, value)
         schemes = [*options.fetch(:schemes)].map(&:to_s)
         begin
           uri = URI.parse(value)
-          unless uri && uri.host && schemes.include?(uri.scheme) && (!options.fetch(:no_local) || uri.host.include?('.'))
-            record.errors.add(attribute, :url, filtered_options(value))
+          validate_suffix = !options.fetch(:public_suffix) || (uri && uri.host && PublicSuffix.valid?(uri.host, :default_rule => nil))
+          validate_no_local = !options.fetch(:no_local) || uri.host.include?('.')
```
Comment (dwilkie) [inline lib/validate_url.rb]: This breaks urls which don't have a host e.g. `www.example.com` because `host` returns `nil`

Comment (zapo) [reply lib/validate_url.rb]: Oh my bad, will PR a hotfix.

**Pull request merged**